syntax = "proto3";

option java_multiple_files = true;
option java_outer_classname = "CappingProto";
option objc_class_prefix = "CAP";

package capping;

// Interface exported by the server.

service Capping {
	rpc Cap(CapInput) returns (CapResult) {}
}

message Mcap {	
	double mcap = 1;
	repeated string components = 2;  // The grouping components
	string ConstituentId = 3;
}

enum NotEnoughComponentsBehaviour {
	NotEnoughComponentsBehaviour_Error = 0;
	NotEnoughComponentsBehaviour_OneOverN = 1;
	NotEnoughComponentsBehaviour_NotApplicable = 2;
}

enum Methodology {
	Methodology_Fixed = 0;
	Methodology_Multi = 1;
	Methodology_Ladder = 2;
	Methodology_Regulatory = 3;  //Not implemented
	Methodology_Exposure = 4;
	Methodology_FactorLimiter = 5;  //Not implemented
	Methodology_CapFloor = 6;	 //Not implemented
}

message LimitInfo {
	double limit = 1;
	optional string limitName = 2;
}


message MethodologyData {
	Methodology methodology = 1;
	repeated LimitInfo limitInfos = 2;
	NotEnoughComponentsBehaviour notEnoughComponentsBehaviour = 4;
}

message CapInput {
	repeated MethodologyData methodologyDatas = 1;
	repeated Mcap parent_mcaps = 2;
	repeated Mcap mcaps = 3;
	bool mcapDecreasingFactors = 4;
}

message CapResult {
    repeated Capfactor capfactors =1;
}

message Capfactor {
    string ConstituentID = 1;
    double factor = 2;
}
